File: go/bob/bob.go
// This is a "stub" file.  It's a little start on your solution.
// It's not a complete solution though; you have to write some code.

// Package bob should have a package comment that summarizes what it's about.
// https://golang.org/doc/effective_go.html#commentary
package bob

import "strings"

// Hey should have a comment documenting it.
func Hey(remark string) string {
	lastChar := remark[len(remark)-1]
	if lastChar == '?' && strings.ToUpper(remark) == remark {
		return "Calm down, I know what I'm doing!"
	}
	if lastChar == '?' {
		return "Sure."
	}
	if strings.ToUpper(remark) == remark {
		return "Whoa, chill out!"
	}
	return "Whatever."
}

File: ruby/bob/bob.rb
# Write your code for the 'Bob' exercise in this file. Make the tests in
# `bob_test.rb` pass.
#
# To get started with TDD, see the `README.md` file in your
# `ruby/bob` directory.

class Bob
  def self.upper?(c)
    c >= 'A' && c <= 'Z'
  end

  def self.question?(remark)
    last = remark.chars.last
    last == '?'
  end

  def self.all_upper?(remark)
    text = remark.gsub(/[^a-zA-Z]/,'')
    return false if text.size.zero?
    text.chars.all? { |c| upper?(c) }
  end

  def self.hey(remark)
    return 'Fine. Be that way!' if remark.strip.size.zero?
    return "Calm down, I know what I'm doing!" if self.question?(remark) && self.all_upper?(remark)
    return "Sure." if self.question?(remark.strip)
    return "Whoa, chill out!" if self.all_upper?(remark)
    'Whatever.'
  end
end
